homeassistant/components/growatt_server/__init__.py

```diff
@@ -19,36 +20,112 @@
 ) -> tuple[list[dict[str, str]], str]:
     """Retrieve the device list for the selected plant."""
     plant_id = config[CONF_PLANT_ID]
 
     # Log in to api and fetch first plant if no plant id is defined.
-    login_response = api.login(config[CONF_USERNAME], config[CONF_PASSWORD])
-    if (
-        not login_response["success"]
-        and login_response["msg"] == LOGIN_INVALID_AUTH_CODE
-    ):
-        raise ConfigEntryError("Username, Password or URL may be incorrect!")
+    try:
+        login_response = api.login(config[CONF_USERNAME], config[CONF_PASSWORD])
+        # DEBUG: Log the actual response structure
+    except Exception as ex:
+        _LOGGER.error("DEBUG - Login response: %s", login_response)
+        raise ConfigEntryError(
+            f"Error communicating with Growatt API during login: {ex}"
+        ) from ex
+
+    if not login_response.get("success"):
+        msg = login_response.get("msg", "Unknown error")
+        _LOGGER.debug("Growatt login failed: %s", msg)
+        if msg == LOGIN_INVALID_AUTH_CODE:
+            raise ConfigEntryAuthFailed("Username, Password or URL may be incorrect!")
+        raise ConfigEntryError(f"Growatt login failed: {msg}")
+
     user_id = login_response["user"]["id"]
+
     if plant_id == DEFAULT_PLANT_ID:
-        plant_info = api.plant_list(user_id)
-        plant_id = plant_info["data"][0]["plantId"]
+        try:
+            plant_info = api.plant_list(user_id)
+        except Exception as ex:
+            raise ConfigEntryError(
+                f"Error communicating with Growatt API during plant list: {ex}"
+            ) from ex
+        if not plant_info or "data" not in plant_info or not plant_info["data"]:
+            raise ConfigEntryError("No plants found for this account.")
+        plant_id = plant_info["data"][0].get("plantId")
+        if not plant_id:
+            raise ConfigEntryError("Plant ID missing in plant info.")
 
     # Get a list of devices for specified plant to add sensors for.
-    devices = api.device_list(plant_id)
+    try:
+        devices = api.device_list(plant_id)
+    except Exception as ex:
+        raise ConfigEntryError(
+            f"Error communicating with Growatt API during device list: {ex}"
+        ) from ex
+
     return devices, plant_id
 
 
+def get_device_list_v1(
+    api, config: Mapping[str, str]
+) -> tuple[list[dict[str, str]], str]:
+    """Device list logic for Open API V1.
+
+    Note: Plant selection (including auto-selection if only one plant exists)
+    is handled in the config flow before this function is called. This function
+    only fetches devices for the already-selected plant_id.
+    """
+    plant_id = config[CONF_PLANT_ID]
+    try:
+        devices_dict = api.device_list(plant_id)
+    except growattServer.GrowattV1ApiError as e:
+        raise ConfigEntryError(
+            f"API error during device list: {e} (Code: {getattr(e, 'error_code', None)}, Message: {getattr(e, 'error_msg', None)})"
```

Comment:
`growattServer.GrowattV1ApiError` always has `error_code` and `error_msg` as property, so we can do without `getattr`